### ISSUE TITLE
Can't find libsyntax from rust sources in external librairies

### ISSUE THREAD

**Issue by amarant** [intellij-rust/intellij-rust#946]
Hi,
I tried the Clippy crates but it seems to be unable to locate type in the libsyntax library.
Steps to reproduce :

1. >git clone https://github.com/Manishearth/rust-clippy.git
2. build with nightly
3. import with intellij rust, try to go to the source of syntax::ast in src/main.rs, in external librairies > Rust the libsyntax with the source is located isn't imported.

Thanks.

**Comment by matklad** [comment]
Yeah, clippy would be tricky, because it uses `#![rustc_private]`, and `syntax::ast` is exactly like this: a crate internal to the compiler which is not supposed to be visible outside. So, it is not present in `cargo metadata` output. 

I think we can handle this with a bit of work though, the same way we deal with standard library itself.

**Comment by CraftSpider** [comment]
Would love to see this implemented, as it's the biggest papercut for exploring/developing on the rustc project that I've run into.

**Comment by disconnect3d** [comment]
Bump. @matklad any chance this gets implemented soon?

**Comment by disconnect3d** [comment]
FWIW this issue can be solved by explicitly adding the private Rust packages needed to the `Cargo.toml`, for example:

```
[dependencies]
rustc_hir = { path = "/nix/store/rqcmifq7iggf50ybmpwx59cf1p90i8ii-rustc-1.49.0/lib/rustlib/rustc-src/rust/compiler/rustc_hir" }
rustc_lint = { path = "/nix/store/rqcmifq7iggf50ybmpwx59cf1p90i8ii-rustc-1.49.0/lib/rustlib/rustc-src/rust/compiler/rustc_lint" }
rustc_middle = { path = "/nix/store/rqcmifq7iggf50ybmpwx59cf1p90i8ii-rustc-1.49.0/lib/rustlib/rustc-src/rust/compiler/rustc_middle" }
rustc_session = { path = "/nix/store/rqcmifq7iggf50ybmpwx59cf1p90i8ii-rustc-1.49.0/lib/rustlib/rustc-src/rust/compiler/rustc_session" }
```

**Comment by CraftSpider** [comment]
My workaround has been to use target dependencies for an invalid target, which ensures those dependencies aren't actually activated by cargo. So same as above, but with `[target.'cfg(target_os = "intellij-hack")'.dependencies]` as the first line instead.

**Comment by disconnect3d** [comment]
@CraftSpider Actually, when I do this and then try to compile via `cargo test` I am getting:
```
error[E0723]: trait bounds other than `Sized` on const fn parameters are unstable
   --> /Users/dc/.cargo/registry/src/github.com-1ecc6299db9ec823/lock_api-0.4.4/src/remutex.rs:224:19
    |
224 | impl<R: RawMutex, G: GetThreadId, T> ReentrantMutex<R, G, T> {
    |                   ^
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn)]` to the crate attributes to enable
```

Adding `#![feature(const_fn)]` to `src/lib.rs` or any other sources doesn't seem to help; do you maybe know how to fix this? It seems the only way to fix this may be changing Rust version? (I may have mismatched libs version vs my Rust compiler version?)

**Comment by ambition-consulting** [comment]
Supporting language libraries is a basic/must-have feature of any IDE. Please prioritize this ticket, or I will have to leave IntelliJ after years of subscribing to Ultimate.
